src/Storages/VirtualColumnUtils.cpp

```diff
@@ -132,6 +147,13 @@ VirtualColumnsDescription getVirtualsForFileLikeStorage(const ColumnsDescription
     add_virtual("_size", makeNullable(std::make_shared<DataTypeUInt64>()));
     add_virtual("_time", makeNullable(std::make_shared<DataTypeDateTime>()));
 
+    for (const auto& path : paths)
+    {
+        auto names = parseVirtualColumnNameFromPath(path);
+        for (const auto& name : names)
+            add_virtual("_" + name, std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()));
```

Comment:
Do we want to use `String` type for all these virtual columns?
For me, ideally we should infer the type of the each value and use inferred type, so for this case:
```
path/key1=42/key2=2020-01-01/key3=[1,2,3]/path
```
We will have `_key1 LowCardinality(Int64), _key2 LowCardinality(Date), _key3 Array(Int64)`.

Reply:
That would require listing all objects before reading, I suppose. Right now, we are doing it lazily, and that's a very nice feature.

We can list all (and make a query potentially slower). Or list a prefix, deduce type from it, and throw an exception if further type conversion did not succeed. Or use strings for everything.

BTW I think that using strings is not very bad.